### ISSUE TITLE
xdg-desktop-portal-cosmic fails due to missing `libclang.so` 

### ISSUE THREAD

**Issue by mainrs** [fsvm88/cosmic-overlay#35]
I think it is the same as #6. Happens both on 9999 and alpha4.

**Comment by mainrs** [comment]
Here some system information:

```console
$ emerge --info
Portage 3.0.66.1 (python 3.12.7-final-0, default/linux/amd64/23.0/desktop, gcc-14, glibc-2.40-r5, 6.6.62-gentoo-dist x86_64)
=================================================================
System uname: Linux-6.6.62-gentoo-dist-x86_64-Intel-R-_Core-TM-_i5-6500_CPU_@_3.20GHz-with-glibc2.40
KiB Mem:    16304052 total,  15784144 free
KiB Swap:   16777212 total,  16777212 free
Timestamp of repository gentoo: Tue, 10 Dec 2024 00:00:00 +0000
Head commit of repository gentoo: 86f85d38c70bde78750cbf27f79e790886ab1910
Head commit of repository cosmic-overlay: fef360d30e6192dd3387ed07ea0f68a0302676b7

Timestamp of repository librewolf: Mon, 09 Dec 2024 22:48:10 +0000
Head commit of repository librewolf: 1c15556d287d3ac6bd933f00bed6aeb2641105cf

Head commit of repository nix-guix: 941bd4ad2d398798fd8903daf76d04a5f74f978d

sh bash 5.2_p37
ld GNU ld (Gentoo 2.43 p3) 2.43.1
distcc 3.4 x86_64-pc-linux-gnu [disabled]
ccache version 4.10.2 [disabled]
app-misc/pax-utils:        1.3.8::gentoo
app-shells/bash:           5.2_p37::gentoo
dev-build/autoconf:        2.72-r1::gentoo
dev-build/automake:        1.16.5-r2::gentoo
dev-build/cmake:           3.30.6::gentoo
dev-build/libtool:         2.4.7-r4::gentoo
dev-build/make:            4.4.1-r100::gentoo
dev-build/meson:           1.5.2::gentoo
dev-lang/perl:             5.40.0::gentoo
dev-lang/python:           3.12.7_p1::gentoo, 3.13.0::gentoo
dev-lang/rust-bin:         1.82.0-r100::gentoo
dev-util/ccache:           4.10.2-r1::gentoo
sys-apps/baselayout:       2.17::gentoo
sys-apps/openrc:           0.55.1::gentoo
sys-apps/sandbox:          2.39::gentoo
sys-devel/binutils:        2.43-r2::gentoo
sys-devel/binutils-config: 5.5.2::gentoo
sys-devel/clang:           18.1.8::gentoo
sys-devel/gcc:             14.2.1_p20241116::gentoo
sys-devel/gcc-config:      2.11::gentoo
sys-devel/llvm:            18.1.8-r1::gentoo, 19.1.4::gentoo
sys-kernel/linux-headers:  6.12::gentoo (virtual/os-headers)
sys-libs/glibc:            2.40-r5::gentoo
Repositories:

gentoo
    location: /var/db/repos/gentoo
    sync-type: rsync
    sync-uri: rsync://rsync.gentoo.org/gentoo-portage
    priority: -1000
    volatile: False
    sync-rsync-verify-jobs: 1
    sync-rsync-extra-opts:
    sync-rsync-verify-metamanifest: yes
    sync-rsync-verify-max-age: 3

cosmic-overlay
    location: /var/db/repos/cosmic-overlay
    sync-type: git
    sync-uri: https://github.com/fsvm88/cosmic-overlay.git
    masters: gentoo
    volatile: False

librewolf
    location: /var/db/repos/librewolf
    sync-type: git
    sync-uri: https://github.com/gentoo-mirror/librewolf.git
    masters: gentoo
    volatile: False

nix-guix
    location: /var/db/repos/nix-guix
    sync-type: git
    sync-uri: https://github.com/trofi/nix-guix-gentoo
    masters: gentoo
    volatile: False

mainrs
    location: /home/me/dev/local/mainrs
    masters: gentoo
    priority: 100
    volatile: True

Binary Repositories:

gentoobinhost
    priority: 1
    sync-uri: https://distfiles.gentoo.org/releases/amd64/binpackages/23.0/x86-64

Installed sets: @base, @base-dev, @base-gui, @base-gui-bin, @de-cosmic, @dev-gentoo
ACCEPT_KEYWORDS="amd64"
ACCEPT_LICENSE="@FREE"
CBUILD="x86_64-pc-linux-gnu"
CFLAGS="-O2 -pipe -march=skylake -mabm --param=l1-cache-line-size=64 --param=l1-cache-size=32 --param=l2-cache-size=6144"
CHOST="x86_64-pc-linux-gnu"
CONFIG_PROTECT="/etc /usr/share/config /usr/share/gnupg/qualified.txt"
CONFIG_PROTECT_MASK="/etc/ca-certificates.conf /etc/dconf /etc/env.d /etc/fonts/fonts.conf /etc/gconf /etc/gentoo-release /etc/sandbox.d"
CXXFLAGS="-O2 -pipe -march=skylake -mabm --param=l1-cache-line-size=64 --param=l1-cache-size=32 --param=l2-cache-size=6144"
DISTDIR="/var/cache/distfiles"
EMERGE_DEFAULT_OPTS="--jobs=2 --load-average=1.8"
ENV_UNSET="CARGO_HOME DBUS_SESSION_BUS_ADDRESS DISPLAY GDK_PIXBUF_MODULE_FILE GOBIN GOPATH PERL5LIB PERL5OPT PERLPREFIX PERL_CORE PERL_MB_OPT PERL_MM_OPT XAUTHORITY XDG_CACHE_HOME XDG_CONFIG_HOME XDG_DATA_HOME XDG_RUNTIME_DIR XDG_STATE_HOME"
FCFLAGS="-O2 -pipe -march=skylake -mabm --param=l1-cache-line-size=64 --param=l1-cache-size=32 --param=l2-cache-size=6144"
FEATURES="assume-digests binpkg-docompress binpkg-dostrip binpkg-logs binpkg-multi-instance buildpkg-live config-protect-if-modified distlocks ebuild-locks fixlafiles ipc-sandbox merge-sync merge-wait multilib-strict network-sandbox news parallel-fetch parallel-install pid-sandbox pkgdir-index-trusted preserve-libs protect-owned qa-unresolved-soname-deps sandbox sfperms strict unknown-features-warn unmerge-logs unmerge-orphans userfetch userpriv usersandbox usersync xattr"
FFLAGS="-O2 -pipe -march=skylake -mabm --param=l1-cache-line-size=64 --param=l1-cache-size=32 --param=l2-cache-size=6144"
GENTOO_MIRRORS="https://ftp.agdsn.de/gentoo     https://ftp.halifax.rwth-aachen.de/gentoo/     https://ftp.uni-stuttgart.de/gentoo-distfiles/"
LANG="en_US.utf8"
LDFLAGS="-Wl,-O1 -Wl,--as-needed -Wl,-z,pack-relative-relocs"
LEX="flex"
MAKEOPTS="-j4 -l5"
PKGDIR="/var/cache/binpkgs"
PORTAGE_CONFIGROOT="/"
PORTAGE_RSYNC_OPTS="--recursive --links --safe-links --perms --times --omit-dir-times --compress --force --whole-file --delete --stats --human-readable --timeout=180 --exclude=/distfiles --exclude=/local --exclude=/packages --exclude=/.git"
PORTAGE_TMPDIR="/var/tmp"
RUSTFLAGS=" -C target-cpu=native"
SHELL="/bin/bash"
USE="X a52 aac acl acpi alsa amd64 bash-completion bluetooth branding bzip2 cairo cdda cdr cet crypt cups dbus dist-kernel dri dts dvd dvdr elogind encode exif flac gdbm gif gpm gtk gui iconv icu ipv6 jpeg kf6compat lcms libnotify libtirpc lto mad mng mp3 mp4 mpeg multilib ncurses nls ogg opengl openmp pam pango pcre pdf png policykit ppds qml qt5 qt6 readline sdl seccomp sound spell ssl startup-notification svg test-rust tiff truetype udev udisks unicode upower usb vorbis vulkan wayland wxwidgets x264 xattr xcb xft xml xv xvid zlib" ABI_X86="64" ADA_TARGET="gcc_12" APACHE2_MODULES="authn_core authz_core socache_shmcb unixd actions alias auth_basic authn_anon authn_dbm authn_file authz_dbm authz_groupfile authz_host authz_owner authz_user autoindex cache cgi cgid dav dav_fs dav_lock deflate dir env expires ext_filter file_cache filter headers include info log_config logio mime mime_magic negotiation rewrite setenvif speling status unique_id userdir usertrack vhost_alias" CALLIGRA_FEATURES="karbon sheets words" COLLECTD_PLUGINS="df interface irq load memory rrdtool swap syslog" CPU_FLAGS_X86="mmx mmxext sse sse2 aes avx avx2 f16c fma3 pclmul popcnt rdrand sse3 sse4_1 sse4_2 ssse3" ELIBC="glibc" GPSD_PROTOCOLS="ashtech aivdm earthmate evermore fv18 garmin garmintxt gpsclock greis isync itrax navcom oceanserver oncore rtcm104v2 rtcm104v3 sirf skytraq superstar2 tsip tripmate tnt ublox" GUILE_SINGLE_TARGET="3-0" GUILE_TARGETS="3-0" INPUT_DEVICES="libinput" KERNEL="linux" L10N="en-US en" LCD_DEVICES="bayrad cfontz glk hd44780 lb216 lcdm001 mtxorb text" LUA_SINGLE_TARGET="lua5-1" LUA_TARGETS="lua5-1" OFFICE_IMPLEMENTATION="libreoffice" PHP_TARGETS="php8-2" POSTGRES_TARGETS="postgres16" PYTHON_SINGLE_TARGET="python3_12" PYTHON_TARGETS="python3_12" RUBY_TARGETS="ruby32" VIDEO_CARDS="intel nvidia" XTABLES_ADDONS="quota2 psd pknock lscan length2 ipv4options ipp2p iface geoip fuzzy condition tarpit sysrq proto logmark ipmark dhcpmac delude chaos account"
Unset:  ADDR2LINE, AR, ARFLAGS, AS, ASFLAGS, CC, CCLD, CONFIG_SHELL, CPP, CPPFLAGS, CTARGET, CXX, CXXFILT, ELFEDIT, EXTRA_ECONF, F77FLAGS, FC, GCOV, GPROF, INSTALL_MASK, LC_ALL, LD, LFLAGS, LIBTOOL, LINGUAS, MAKE, MAKEFLAGS, NM, OBJCOPY, OBJDUMP, PORTAGE_BINHOST, PORTAGE_BUNZIP2_COMMAND, PORTAGE_COMPRESS, PORTAGE_COMPRESS_FLAGS, PORTAGE_RSYNC_EXTRA_OPTS, PYTHONPATH, RANLIB, READELF, SIZE, STRINGS, STRIP, YACC, YFLAGS
```

**Comment by fsvm88** [comment]
Just pushed a commit that should fix it. Can you test? (mentioning the issue in the git commit automatically closes it - feel free to reopen if not working)

**Comment by enx1998** [comment]
Hi @fsvm88 
I have exactly the same issue: `cosmic-de/xdg-desktop-portal-cosmic-1.0.0_alpha4` fail to emerge with error:
> thread 'main' panicked at /var/tmp/portage/cosmic-de/xdg-desktop-portal-cosmic-1.0.0_alpha4/work/cargo_home/gentoo/bindgen/lib.rs:622:31:
  Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"
  
Do you think is this issue fixable?
Thanks ENx

**Comment by mainrs** [comment]
It should be fixed already. Make sure to delete the local copy that portage creates. I think it was under `/var/tmp/portage/cosmic-de/xdg-desktop-portal-cosmic`. Just `rm -rf` it.

**Comment by simonvanderveldt** [comment]
Just ran into what seems to be the same issue as well
```
     Running `/var/tmp/portage/cosmic-de/xdg-desktop-portal-cosmic-9999/work/xdg-desktop-portal-cosmic-9999/target/release/build/libspa-sys-3a0fe3b257f7a74c/build-script-build`
error: failed to run custom build command for `libspa-sys v0.8.0 (https://gitlab.freedesktop.org/pipewire/pipewire-rs#86df3919)`

Caused by:
  process didn't exit successfully: `/var/tmp/portage/cosmic-de/xdg-desktop-portal-cosmic-9999/work/xdg-desktop-portal-cosmic-9999/target/release/build/libspa-sys-3a0fe3b257f7a74c/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=LIBPIPEWIRE_0.3_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=SYSROOT
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=LIBSPA_0.2_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=SYSROOT
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:rustc-link-lib=pipewire-0.3
  cargo:include=/usr/include/pipewire-0.3:/usr/include/spa-0.2:/usr/include/spa-0.2
  cargo:rerun-if-env-changed=SYSTEM_DEPS_BUILD_INTERNAL
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LINK
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_LIB
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_LIB_FRAMEWORK
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_SEARCH_NATIVE
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_SEARCH_FRAMEWORK
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_INCLUDE
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_BUILD_INTERNAL
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBSPA_LINK
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_LIB
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_LIB_FRAMEWORK
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_SEARCH_NATIVE
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_SEARCH_FRAMEWORK
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_INCLUDE
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_BUILD_INTERNAL
  cargo:rerun-if-env-changed=SYSTEM_DEPS_LIBPIPEWIRE_LINK

  cargo:rustc-cfg=system_deps_have_libpipewire
  cargo:rustc-cfg=system_deps_have_libspa
  cargo:rerun-if-changed=wrapper.h
  cargo:rerun-if-env-changed=TARGET
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS
  cargo:rerun-if-changed=wrapper.h

  --- stderr
  thread 'main' panicked at /var/tmp/portage/cosmic-de/xdg-desktop-portal-cosmic-9999/work/cargo_home/gentoo/bindgen/lib.rs:622:31:
  Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"
```

LLVM (18.1.8-r6) is installed (The package is now called `llvm-core/clang` btw, see https://github.com/gentoo/gentoo/commit/3ee411f1324ad2b881555d1bec5f943f1a7f28a1)

**Comment by fsvm88** [comment]
Hey! Thanks for the update.

Is this perhaps the first emerge on a clean/stage3 system that did not have clang installed?

What I mean is:
1. Extract stage3
2. Add overlay
3. `emerge cosmic-de/cosmic-meta`

The only reason I can think of, if the package is installed at the time of failure, is that it wasn't installed before emerging, thus `/etc/env.d/60llvm-9980` did not exist, and was never sourced by an `env-update && source /etc/profile`. The search path for llvm/clang libraries would then not be present in $LDPATH.

**Comment by picarica** [comment]
ran into this issue again today, is this still a issue? i am emerging version 1.0.0_alpha4, should i merge version 9999? i am unable to merge other packages in that version, tho, is there some temporary fix ?

**Comment by simonvanderveldt** [comment]
> Is this perhaps the first emerge on a clean/stage3 system that did not have clang/llvm installed before the merge?

No, unfortunately not, this was on my desktop which has been using LLVM for a long while because it's needed for my GPU.

I'll try to do some digging later this week.

@picarica I had the same issue with the live ebuild, but you can always try.

**Comment by fsvm88** [comment]
I see. Thanks for updates.

I've managed to do some more digging, and I think I've found the issue.

In @mainrs output above I noticed that clang 18 is installed and used, but llvm 18 AND llvm 19 are present as well.

According to the logic in LLVM's ebuild, a new drop-in env file is created at `/etc/env.d/60llvm-$((9999 - LLVM_MAJOR))`, which means that *higher* llvm versions appear *earlier* in PATH (e.g. /bin:/sbin:../llvm/19/bin:/.../llvm/18/bin:...).

So when `libclang-sys` (dep of `pipewire-rs` -> `libspa-sys`) auto-detects the llvm version, it just calls the first `llvm-config` found in PATH (because atm we don't override it/manage it). At least in @mainrs's case this means that `llvm-config --libdir` returns `/usr/lib/llvm/19/lib64`, but he doesn't seem to have clang-19 installed (otherwise I suspect that by the same logic, `emerge --info` would have picked clang-19), thus, effectively, the error is real and correct: there is no *libclang.so* available on the system that matches the `llvm-config --libdir` auto-detected version.


TL; DR: can you please try explicitly override PATH in your `emerge` invocation and test if this fixes it?

I believe something like this should allow compiling to succeed for now:
```
PATH=/usr/lib/llvm/18/bin:$PATH emerge -1 xdg-desktop-portal-cosmic
```

If your testing is positive, I can find a way to supply this explicitly, or at least require a specific version of clang instead. I can't test the reverse because I already got rid of all llvm-18 deps on my systems :)

Let me know :)

**Comment by simonvanderveldt** [comment]
@fsvm88 I think you were right, somehow I did have LLVM 18 and 19 installed (which installed the file in `/etc/env.d`) but had only Clang 18 install, Clang 19 was missing, not really sure why. After emerging Clang 19 compilation of xdg-desktop-portal-cosmic 9999 succeeded without issues.
Tbh I believe this was an issue on my side, probably a not fully finished update/emerge that caused this, so I don't think there's anything that would need fixing. Not sure about the other reports though.

**Comment by fsvm88** [comment]
Thanks for testing!

> Tbh I believe this was an issue on my side, probably a not fully finished update/emerge that caused this, so I don't think there's anything that would need fixing. Not sure about the other reports though.

I don't think so -> some packages only require LLVM and have no need for clang, so only LLVM is pulled in.

Some other packages (e.g.: firefox[clang]) normally don't fix the version required for building, or fix it to the "best" LLVM_SLOT available (so if you force LLVM_SLOT=18 in make.conf, or you are on stable and the package has no LLVM_SLOT=19 compat, you only get clang 18).

So you can end up with a system for which some packages have LLVM_SLOT="18 19", but where the only package requiring clang has LLVM_SLOT=18 (or a similar situation). If you don't update too often, or you only use stable, the next clang version could lag behind for a while, esp. if you don't update with `-N`. I'm guessing the other reports are in the same condition.

I guess I'll maybe implement LLVM_SLOT, or perhaps resort to specifying matching llvm and clang in the dependencies. I'll think about it tomorrow, for today I'm spent :)

**Comment by fsvm88** [comment]
It should be fixed with the referenced commit.

Using llvm-r1 eclass, affected ebuilds should automatically select the proper slot based on what is already available, automatically depend on newer versions if/when required, and prepend the proper path at all times.

Let me know if this fixes it for good.
